### PR TITLE
Controlled accounts

### PR DESCRIPTION
**Pull request by wmurphyrd**
- Implement OAuth2 JWT exchange for 2-legged OAuth service accounts
- Include pending, outgoing requests in friends api response to correctly populate friends lists without relying on local state
- Allow service account JWTs to authenticate user creation requests when public registration is disabled

### CODE REVIEW

File: src/authdb.js [hunk 1]
```diff
@@ -132,7 +168,9 @@ module.exports = {
   },
   async createUser (username, password, email) {
     const user = { username }
-    user.passwordHash = await bcrypt.hash(password, saltRounds)
+    if (password) {
+      user.passwordHash = await bcrypt.hash(password, saltRounds)
+    }
```
Comment (wmurphyrd) [inline src/authdb.js:173]: This allows creating user without password, [existing code](https://github.com/immers-space/immers/blob/controlled-accounts/src/authdb.js#L64) in `validateUser` already checks and rejects authentication requests when the user has no saved password

**Comment by wmurphyrd** [comment]
It just hit me the way I should have done the user registration override.
1. Leave the existing route and settings methods unchanged
2. Put another POST /auth/user before the original that checks for the presence of a bearer token, authorizes with it if so,  and passes onto the original route with next('route') if not

Gonna draft this PR and go back and do it that way

File: doc/ControlledAccounts.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+Your admin private key must be kept private on your server and never sent to the client.
+Sign a JWT and send it as a OAuth2 token exchange to retrieve a user access token.
+
+```js
+const { readFileSync } = require('fs')
+const jwt = require('jsonwebtoken')
+const axios = require('axios') // any request library will do
+const immersAdminPrivateKey = readFileSync('immersAdminPrivateKey.pem')
+const auth = jwt.sign(
```
Comment (quinn-madson) [inline doc/ControlledAccounts.md:69]: I think this is supposed to be: **oAuthJwt** instead of **auth**.

File: doc/ControlledAccounts.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  immersAdminPrivateKey,
+  {
+    algorithm: "RS256",
+    issuer: `https://yourDomain.com/o/immer`,
+    audience: `https://yourDomain.com/o/immer`,
+    expiresIn: "1h",
+    subject: "user[yourdomain.dom]" // the user that you will login as
+  }
+);
+const params = new url.URLSearchParams({
```
Comment (quinn-madson) [inline doc/ControlledAccounts.md:83]: I don't think params is being used here.

**Review by quinn-madson — CHANGES_REQUESTED** [review]

**Review by quinn-madson — APPROVED** [review]

**Pull request merged**